pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -367,3 +379,42 @@ func (h *VMController) ManageOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine
 
 	return h.OnVMRemove(vm)
 }
+
+func (h *VMController) SetHaltIfInsufficientResourceQuota(_ string, vm *kubevirtv1.VirtualMachine) (*kubevirtv1.VirtualMachine, error) {
+	if vm == nil || vm.Spec.Template == nil {
+		return vm, nil
+	}
+	if exist, err := h.vmrCalculator.ExistsVMPod(vm.Namespace, vm.Name); err != nil || exist {
+		return vm, err
+	}
+
+	if strategy, err := vm.RunStrategy(); err != nil || strategy == kubevirtv1.RunStrategyHalted {
+		return vm, err
+	}
+
+	if err := h.vmrCalculator.CheckVMLimits(vm); err == nil {
+		return vm, nil
+	} else if !resourcequota2.IsInsufficientResourceError(err) {
+		return nil, err
+	}
+
+	runStrategy, err := vm.RunStrategy()
+	if err != nil || runStrategy == kubevirtv1.RunStrategyHalted {
+		return vm, err
+	}
+
+	if err := h.stopVM(vm); err != nil {
+		return nil, err
+	}
+
+	h.recorder.Eventf(vm, corev1.EventTypeWarning, "InsufficientResourceQuota", "Set runStrategy to %s due to %s", kubevirtv1.RunStrategyHalted, err)
```

Comment:
last arg of `Set runStrategy to %s due to %s` should not be err. err always nil here